jax/_src/api.py

```diff
@@ -333,15 +335,67 @@ def jit(
     DeviceArray([   0,    1,  256, 6561], dtype=int32)
   """
   if FLAGS.experimental_cpp_jit and not config.jax_dynamic_shapes:
-    return _cpp_jit(fun, static_argnums, static_argnames, device, backend,
+    return _jit("cpp", fun, static_argnums, static_argnames, device, backend,
                     donate_argnums, inline, keep_unused)
   else:
-    return _python_jit(fun, static_argnums, static_argnames, device, backend,
+    return _jit("python", fun, static_argnums, static_argnames, device, backend,
                        donate_argnums, inline, keep_unused, abstracted_axes)
 
+def _jit(
+    jit_implementation: Literal["cpp", "python", None],
+    fun: Callable,
+    static_argnums: Union[int, Iterable[int], None] = None,
+    static_argnames: Union[str, Iterable[str], None] = None,
+    device: Optional[xc.Device] = None,
+    backend: Optional[str] = None,
+    donate_argnums: Union[int, Iterable[int]] = (),
+    inline: bool = False,
+    keep_unused: bool = False,
+    abstracted_axes: Optional[Any] = None,
+  ) -> stages.Wrapped:
+  # Implemements common logic between CPP and Python backends
+  _check_callable(fun)
+
+  # Infer argnums and argnames according to docstring
+  static_argnums, static_argnames = _infer_argnums_and_argnames(
+      fun, static_argnums, static_argnames)
+
+  # Coerce input
+  donate_argnums = _ensure_index_tuple(donate_argnums)
+
+  # Validation
+  validate_argnums(fun, static_argnums, "static_argnums")
```

Comment:
Since making validation stricter is liable to break user code, we may have to start by reporting these as warnings not errors.

Let me try running our internal presubmits with this PR, which include a lot of downstream JAX code, to get a sense of how bad the problem will be.

Reply:
That might be a good idea. On the other hand, invalidation only happens when called with genuinely invalid arguments and is very easily fixed. Could be frustrating if a library causes the error.

I don't know what the timeline for iterating on minor versions. My opinion would be that given JAX is still young and experimental, it would be fine to implement this as an exception rather than a warning if you issue a new minor version (or wait with merging this until a new minor version is appropriate). Given JAX is still 0.x.y, this would comply with SemVer afaik.

Reply:
It's not looking too bad. I found that many tests broke because of this bug:
https://github.com/google/trajax/blob/67fd5ed4867914f1ce1ab78819551c12206a7773/trajax/optimizers.py#L753
but that was the main cause of most failures. But there were a handful of other bugs here and there.

In general, our policy is: https://jax.readthedocs.io/en/latest/api_compatibility.html

Reply:
Happy to change over to warnings, if you'd prefer. Given `trajax` is Google too, maybe it would be sufficient to to fix [google/trajax@67fd5ed/trajax/optimizers.py#L753](https://github.com/google/trajax/blob/67fd5ed4867914f1ce1ab78819551c12206a7773/trajax/optimizers.py#L753) and push a `trajax` release and then not change to warnings?

Reply:
Will do this later tonight

Reply:
Done.